### PR TITLE
Adding FetchStatus Method

### PR DESCRIPTION
**Pull request by ahmedbodi**
Just a draft yet! Don't merge!

Initial Work to the base exchange

### CODE REVIEW

**Comment by kroitor** [comment]
Thx! One thing to note: base exchange files are not transpiled, so we need to implement it in JS, Py and PHP.

**Comment by ahmedbodi** [comment]
gotcha, will be sure to implement in all of them

**Comment by kroitor** [comment]
Awesome, thx so much again! )

**Comment by kroitor** [comment]
@ahmedbodi jfyi, in JS we can assign the status to this dynamically, but in Py and PHP it has to be predeclared as a class property before populating it in `describe()`, so we have to add it in two places to each of the `exchange.py` and `Exchange.php`. Just to save you some time, hope it helps.

**Comment by ahmedbodi** [comment]
@kroitor do you know which exchanges have a status endpoint so i can add them?

**Comment by kroitor** [comment]
@ahmedbodi 

- https://github.com/ccxt/ccxt/blob/master/js/binance.js#L89
- https://github.com/ccxt/ccxt/blob/master/js/cobinhood.js#L60
- https://github.com/ccxt/ccxt/blob/master/js/mercado.js#L71

Also, if you look here:

- https://github.com/ccxt/ccxt/blob/master/js/bittrex.js#L83
- https://github.com/ccxt/ccxt/blob/master/js/gdax.js#L69
- https://github.com/ccxt/ccxt/blob/master/js/gdax.js#L421

We can _"emulate"_ the `fetchStatus` `updated` field with `fetchTime`... The logic behind it could be:" _if the exchange responded to that request, it probably worked at that timestamp"_ . However, this has to be investigated, because the Time endpoint can still work when the exchange is down.

**Comment by ahmedbodi** [comment]
Yep, we could use that, it just wouldnt be guaranteed to always be right then. But it should atleast help people who are trying to connect to dead exchanges

**Comment by ahmedbodi** [comment]
i'd say if we are going to use the ping endpoint then we should just keep the status as emulated since we dont have the full picture of the exchange status

**Comment by kroitor** [comment]
> i'd say if we are going to use the ping endpoint then we should just keep the status as emulated since we dont have the full picture of the exchange status

Yep, we could change the base class implementation:

```JavaScript
async fetchStatus (params = {}) {
    if (this.has['fetchTime']) {
        const time = await this.fetchTime (params)
        this.status['updated'] = time
    }
    return this.status
}
```

**Comment by alexeymarunin** [comment]
For Binance:

```js
const binance = new ccxt.binance()
const systemStatus = await binance.wapiGetSystemStatus()
const status = systemStatus.status === 0 ? 'ok' : 'maintenance'
```

See [WAPI System Status](https://github.com/binance-exchange/binance-official-api-docs/blob/master/wapi-api.md#system-status-system)

**Comment by kroitor** [comment]
@alexeymarunin yep, it's on the way as was discussed here: https://github.com/ccxt/ccxt/issues/5285 ) Thx! )

**Comment by ahmedbodi** [comment]
Included FYB-SG to fix #5316

File: js/fybsg.js [hunk 1]
```diff
@@ -24,6 +24,12 @@ module.exports = class fybsg extends fybse {
             'markets': {
                 'BTC/SGD': { 'id': 'SGD', 'symbol': 'BTC/SGD', 'base': 'BTC', 'quote': 'SGD' },
             },
+            'status': {
+                'status': 'shutdown',
```
Comment (kroitor) [inline js/fybsg.js]: Yeah! I'll review it once again thoughtfully, and hope to merge it in the coming days.

Comment (ahmedbodi) [reply js/fybsg.js]: yep sure

**Comment by ahmedbodi** [comment]
@kroitor any other exchanges?

**Comment by kroitor** [comment]
@ahmedbodi i suggest to keep it compact, so that i could merge it soon without too many conflicts, and we will then add the rest of them one by one in the background.

**Comment by ahmedbodi** [comment]
OK sounds good to me, so i'll leave it here since it has an example of fetching the status and another of hardcoded

**Comment by ahmedbodi** [comment]
Also, do we only mark exchanges that are completely not working as errored, or do we include ones that have some failing methods like #5164

**Comment by alexeymarunin** [comment]
Is there documentation for fetchStatus()? I mean how many statuses (`error`, `maintenance`, `shutdown`, `ok`, etc) and what are they mean. Also it helps developers (example, for me) to add fetchStatus() to other exchanges

**Comment by kroitor** [comment]
@alexeymarunin not yet, this is something i'm going to add to this PR before merging. Will document all the statuses and the usage.

**Comment by ahmedbodi** [comment]
@kroitor anything else needed before it can be merged?

**Comment by kroitor** [comment]
@ahmedbodi nope, thank you so much! I'll add the list of all the statuses to your edits to the docs and will merge it shortly (hopefully, today).

**Comment by kroitor** [comment]
@ahmedbodi pardon again for the delays on my side, will get to it asap.

File: js/test/test.js [hunk 2]
```diff
@@ -227,6 +227,7 @@ let testExchange = async exchange => {
 
     await tests['fetchFundingFees']  (exchange)
     await tests['fetchTradingFees']  (exchange)
+    await tests['fetchStatus'] (exchange)
```
Comment (kroitor) [inline js/test/test.js]: After thinking more about it, we should not really try testing the exchange, if `fetchStatus` is true and the `status` is not `ok` ) But I'll merge it this way for now, and will rework the test a bit later.

**Comment by ahmedbodi** [comment]
Cheers @kroitor

**Comment by kroitor** [comment]
@ahmedbodi thx for your help!

**Comment by ahmedbodi** [comment]
np

**Comment by alexeymarunin** [comment]
@kroitor Could you explain what's the difference between `has: fetchStatus: true` and `has: fetchStatus: 'emulated'`? I'm not quite understand

**Comment by kroitor** [comment]
@alexeymarunin if the exchange has a standalone endpoint that would return the status of the exchange (whether it is online an working, on maintenance, shutdown or whatever). If the exchange in question has such an endpoint, we override `fetchStatus` and code the parsing logic there – that's the `has['fetchStatus'] = true` case. If the exchange does not have it, then it falls back to the default base method, that fetches time from the exchange, instead of status – that the `has['fetchStatus'] = 'emulated'` case.

**Pull request merged**